### PR TITLE
Add sequence numbers, bounded queue support, named subscribers.

### PR DESCRIPTION
**Pull request by msebek**
Add sequence numbers, so Subscribers can know when messages are being dropped. 

Add bounded queue support, so messages can be dropped if messages are stuck aging in the queue

Add names for subscribers, to make it easier to figure out who started what threads.

### CODE REVIEW

**Comment by msebek** [comment]
Note that the GUI subscribers have not yet been changed to limit their queue length to 1.

File: real_time/surface_src/java_src/BuggyRos/src/test/java/BoundedQueueNumberSink.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.roboclub.robobuggy.ros.Message;
+import com.roboclub.robobuggy.ros.MessageListener;
+import com.roboclub.robobuggy.ros.Node;
+import com.roboclub.robobuggy.ros.Subscriber;
+
+public class BoundedQueueNumberSink implements Node {
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/BuggyRos/src/test/java/BoundedQueueNumberSink.java]: Codacy Issue found: [All classes and interfaces must belong to a named package](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2362871859/issues/source?bid=2938665&fileBranchId=3220358#l9)

File: real_time/surface_src/java_src/BuggyRos/src/test/java/BoundedQueueNumberSink.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+	private Subscriber s;
+	private Semaphore sem = new Semaphore(1);
+	private AtomicInteger count = new AtomicInteger(1);
+
+	public BoundedQueueNumberSink(String topicName, int numMessages) {
+		this(topicName, numMessages, Integer.MAX_VALUE);
+	}
+	
+	public BoundedQueueNumberSink(String topicName, int numMessages, int maxMessageQueueLength) {
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/BuggyRos/src/test/java/BoundedQueueNumberSink.java]: Codacy Issue found: [Avoid excessively long variable names like maxMessageQueueLength](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2362871859/issues/source?bid=2938665&fileBranchId=3220358#l19)

File: real_time/surface_src/java_src/BuggyRos/src/test/java/UnboundedMessageQueueTest.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import org.junit.Test;
+
+public class UnboundedMessageQueueTest {
+
+	@Before
+	public void setUp() throws Exception {
+	}
+
+	@After
+	public void tearDown() throws Exception {
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/BuggyRos/src/test/java/UnboundedMessageQueueTest.java]: Codacy Issue found: [Document empty method body](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2362871863/issues/source?bid=2938665&fileBranchId=3220358#l14)

File: real_time/surface_src/java_src/BuggyRos/src/test/java/UnboundedMessageQueueTest.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
+import static org.junit.Assert.fail;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class UnboundedMessageQueueTest {
+
+	@Before
+	public void setUp() throws Exception {
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/BuggyRos/src/test/java/UnboundedMessageQueueTest.java]: Codacy Issue found: [Document empty method body](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2362871863/issues/source?bid=2938665&fileBranchId=3220358#l10)

File: real_time/surface_src/java_src/BuggyRos/src/test/java/UnboundedMessageQueueTest.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
+import static org.junit.Assert.fail;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class UnboundedMessageQueueTest {
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/BuggyRos/src/test/java/UnboundedMessageQueueTest.java]: Codacy Issue found: [All classes and interfaces must belong to a named package](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2362871863/issues/source?bid=2938665&fileBranchId=3220358#l7)

Comment (TrevorDecker) [inline real_time/surface_src/java_src/BuggyRos/src/test/java/UnboundedMessageQueueTest.java]: Codacy Issue found: [Each class should declare at least one constructor](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2362871863/issues/source?bid=2938665&fileBranchId=3220358#l7)

File: real_time/surface_src/java_src/BuggyRos/src/main/java/com/roboclub/robobuggy/ros/internal/MessageServer.java [hunk 6]
```diff
@@ -75,9 +77,32 @@ public void run() {
...
+
+	
+	/*
+	 * The StatisticsThread 
+	 */
+	private class StatisticsThread implements Runnable {
+		// Interesting things: how many messages are in the queue, 
+		//					   how many messages processed since last time.
+		
+		public StatisticsThread(int pollingPeriodInMs) {
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/BuggyRos/src/main/java/com/roboclub/robobuggy/ros/internal/MessageServer.java]: Codacy Issue found: [Avoid unused constructor parameters such as 'pollingPeriodInMs'.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2362871856/issues/source?bid=2938665&fileBranchId=3220358#l95)

File: real_time/surface_src/java_src/BuggyRos/src/main/java/com/roboclub/robobuggy/ros/internal/MessageServer.java [hunk 7]
```diff
@@ -36,6 +37,8 @@
 
 	// Contains all messages that threads want sent.
 	private LinkedBlockingQueue<Map.Entry<String, Message>> inbox = new LinkedBlockingQueue<Map.Entry<String, Message>>();
+
+	private int messagesProcessed = 0;
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/BuggyRos/src/main/java/com/roboclub/robobuggy/ros/internal/MessageServer.java]: Codacy Issue found: [Avoid unused private fields such as 'messagesProcessed'.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2362871856/issues/source?bid=2938665&fileBranchId=3220358#l41)

File: real_time/surface_src/java_src/BuggyRos/src/main/java/com/roboclub/robobuggy/ros/internal/MessageServer.java (hunk 6 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/BuggyRos/src/main/java/com/roboclub/robobuggy/ros/internal/MessageServer.java]: Codacy Issue found: [Document empty constructor](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2362871856/issues/source?bid=2938665&fileBranchId=3220358#l95)

File: real_time/surface_src/java_src/BuggyRos/src/main/java/com/roboclub/robobuggy/ros/internal/MessageServer.java [hunk 8]
```diff
@@ -75,9 +77,32 @@ public void run() {
...
+	private class StatisticsThread implements Runnable {
+		// Interesting things: how many messages are in the queue, 
+		//					   how many messages processed since last time.
+		
+		public StatisticsThread(int pollingPeriodInMs) {
+			
+		}
+		
+		@Override
+		public void run() {
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/BuggyRos/src/main/java/com/roboclub/robobuggy/ros/internal/MessageServer.java]: Codacy Issue found: [Document empty method body](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2362871856/issues/source?bid=2938665&fileBranchId=3220358#l100)

**Comment by TrevorDecker** [comment]
I messed up so this is moving to #409, anyways I approve it

**Pull request closed without merging**